falcon/util/mediatypes.py

```diff
@@ -88,3 +92,221 @@ def parse_header(line: str) -> Tuple[str, Dict[str, str]]:
+
+        return cls(main_type, subtype, quality, params)
+
+    def match_score(self, media_type: _MediaType) -> Tuple[int, int, int, int, float]:
+        if self.main_type == '*' or media_type.main_type == '*':
+            main_matches = 0
+        elif self.main_type != media_type.main_type:
+            return self._NOT_MATCHING
+        else:
+            main_matches = 1
+
+        if self.subtype == '*' or media_type.subtype == '*':
+            sub_matches = 0
+        elif self.subtype != media_type.subtype:
+            return self._NOT_MATCHING
+        else:
+            sub_matches = 1
+
+        mr_pnames = frozenset(self.params)
+        mt_pnames = frozenset(media_type.params)
+
+        exact_match = 0 if mr_pnames.symmetric_difference(mt_pnames) else 1
+
+        matching = mr_pnames & mt_pnames
+        for pname in matching:
+            if self.params[pname] != media_type.params[pname]:
+                return self._NOT_MATCHING
+
+        return (main_matches, sub_matches, exact_match, len(matching), self.quality)
+
+    def __repr__(self) -> str:
+        q = self.quality
+        return f'MediaRange<{self.main_type}/{self.subtype}; {q=}; {self.params}>'
+
+
+# PERF(vytas): It is possible to cache a classmethod too, but the invocation is
+#   less efficient, especially in the case of a cache hit.
+_parse_media_type = functools.lru_cache(_MediaType.parse)
+_parse_media_range = functools.lru_cache(_MediaRange.parse)
+
+
+@functools.lru_cache()
+def _parse_media_ranges(header: str) -> Tuple[_MediaRange, ...]:
+    return tuple(_MediaRange.parse(media_range) for media_range in header.split(','))
+
+
+@functools.lru_cache()
+def quality(media_type: str, header: str) -> float:
+    """Get quality of the most specific matching media range.
+
+    Media-ranges are parsed from the provided `header` value according to
+    RFC 9110, Section 12.5.1 (the ``Accept`` header).
+
+    The provided `media_type` is matched against each of the parsed media
+    ranges, and the fitness of each match is assessed as follows
+    (in the decreasing priority list of criteria):
+
+    1. Do the main types (as in ``type/subtype``) match?
+
+       The types must either match exactly, or as wildcard (``*``).
+       The matches involving a wildcard are prioritized lower.
+
+    2. Do the subtypes (as in ``type/subtype``) match?
+
+       The subtypes must either match exactly, or as wildcard (``*``).
+       The matches involving a wildcard are prioritized lower.
+
+    3. Do the parameters match exactly?
+
+       If all the parameter names and values (if any) between the media range
+       and media type match exactly, such match is prioritized higher than
+       matches involving extraneous parameters on either side.
+
+       Note that if parameter names match, the corresponding values must also
+       be equal, or the provided media type is considered not to match the
+       media range in question at all.
```

Comment:
The current behaviour (which I think is problematic) is that it is not a big deal if the values don't match.

But I was reading RFCs, and logically thinking, if one specifies `type/subtype; param=value` in `Accept` or in media handler, it means that this specific media range or media handler applies only when `param=value`. So I don't think it makes sense to allow mismatching values, or maybe we should heavily penalize them?

I can add a note that if this beheaviour is not desirable, one should always also add a handler for the generic media range without parameters.

Reply:
yes, but it depends what you want to empathize. One could say that `type/subtype` do match it's only an option of those that don't.
For example `application/json; indent=2`. If we support `application/json; indent=4` once could argue that it's fine to match that mime parse.

Here either case we choose is likely wrong some times, but I feel like we could place this a the least important weight subtracting one each time a parameter is wrong, so that we get in order of same q type with no mismatch parameters or if there are miss-matched ones the one with least mis-match

Reply:
If a function can only handle JSON with `indent=2`, it is likely to explode with `indent=4`, I would say it is a good example for matching params.

I think it is also in the spirit of RFC 9110 that if provided, parameters must match exactly.

I agree with you in general, that since the RFCs don't specify the behaviour exactly, there will always be cases for and against. 
For instance, `charset=UTF-8` and `charset=utf8` currently wouldn't match, although they should. That said, charsets are a very good example of why parameters must match otherwise -- if a function accepts UTF-8, it is very likely gonna explode or malfunction if you submit Latin-1, or even worse, UTF-16.

Reply:
Discussed this more in depth on Gitter, and while there is no clear right and wrong here, it is very unlikely to affect anyone.
We can address complaints if they come.
I will add a note on possible workarounds.